src/pqsort.c

```diff
@@ -62,7 +62,7 @@ static inline void	 swapfunc (char *, char *, size_t, int);
         } while (--i > 0);				\
 }
 
-#define SWAPINIT(a, es) swaptype = ((char *)a - (char *)0) % sizeof(long) || \
+#define SWAPINIT(a, es) swaptype = (unsigned long)a % sizeof(long) || \
```

Comment:
i'm having hard time to understand why the original code did `((char *)a - (char *)0)`.
i guess it's just an awkward way to convert the pointer into integer?
i'm guessing the right way is to actually use `uintptr_t`?
```suggestion
#define SWAPINIT(a, es) swaptype = (uintptr_t)a % sizeof(long) || \
```

Reply:
I saw that a few times in older code, I couldn't find an explanation too. Back then, `uintptr_t` didn't exist (pre C99) and casting pointer to integer type is undefined behavior. Looks like some people thought subtracting NULL pointer is to way to generate a number from a pointer value. It's still undefined behavior though. Maybe some developers interpreted that as standard compatible way back then. Looks like that code is from ~30 years ago, maybe some corner cases were not clear at that time. 30 years after, still many things are not clear whether they are undefined behavior or not :)

Anyway, I agree with you, I think `uintptr_t` is the correct solution.